env/env.cc

```diff
@@ -489,10 +729,8 @@ const std::shared_ptr<FileSystem>& Env::GetFileSystem() const {
   return file_system_;
 }
 
-#ifdef OS_WIN
-std::unique_ptr<Env> NewCompositeEnv(std::shared_ptr<FileSystem> fs) {
+std::unique_ptr<Env> NewCompositeEnv(const std::shared_ptr<FileSystem>& fs) {
```

Comment:
Can we provide a way to create a PosixEnv instance with a custom FileSystem?

Reply:
I do not think that is desirable for a few reasons.  First, it greatly complicates the destruction of the PosixEnv code.  Now instead of a PosixEnv being a singleton, there could potentially be many instances of them, making the destruction of things like the thread pool more complex.  Second, there is no interface to request a "Env::Default" with anything except the Default FileSystem.  Adding an API Env::Default(const std::shared_ptr<FileSystem>& fs) is potentially complicated and error-prone -- What happens if the method is called with different "fs" arguments?

In a future PR, I would like to split the "time/clock" functions out of the Env into a SystemClock class, comparable to what was done with the FileSystem.  This will greatly reduce the number of Env classes there are (for testing) and also clean up some of the code.  Adding more ways of creating a "Default" would greatly complicate that work.

What is the use case where a PosixEnv with a Custom FileSystem is important that cannot be solved efficiently in the current proposal?

Reply:
See my latest comment below. 

```NewCompositeEnv``` is the interface to request a ```Env::Default``` with a custom FileSystem.  Why do we need ```Env::Default(const std::shared_ptr& fs)```? And the current ```NewCompositeEnv``` implementation in env_posix.cc already does what I described - create a new PosixEnv instance with the global shared thread pool, and a custom FileSystem. The destruction of the thread pool is handled properly in the presence of multiple PosixEnv instances. We just need something similar for Windows. If its difficult to do in Windows, maybe the Windows implementation of ```NewCompositeEnv``` can do ```new CompositeEnvWrapper```. I don't see why we need to add some inefficiency to the Posix implementation.